Core/Core/Planner/CalendarFilter/ViewModel/CalendarFilterViewModel.swift

```diff
@@ -0,0 +1,183 @@
+                allContexts.append(contentsOf: groupFilters.map { $0.context })
+                allContexts.appendUnwrapped(userFilter?.context)
+                return (allContexts, selectedContexts.isEmpty)
+            }
+            .flatMap { [interactor] (contexts, isSelect) in
+                interactor.updateFilteredContexts(contexts, isSelected: isSelect)
+            }
+            .sink()
+            .store(in: &subscriptions)
+    }
+
+    private func handleDoneButtonTap() {
+        didTapDoneButton
+            .sink { [router, didDismissPicker] host in
+                router.dismiss(host, completion: didDismissPicker)
+            }
+            .store(in: &subscriptions)
+    }
+
+    private func observeDataChanges() {
+        interactor
+            .selectedContexts
+            .assign(to: \.selectedContexts, on: self, ownership: .weak)
+            .store(in: &subscriptions)
+
+        Publishers.CombineLatest(
+            interactor.selectedContexts,
+            interactor.filterCountLimit
+        )
+            .map { (selectedContexts, filterCountLimit) in
+                if filterCountLimit != .unlimited, selectedContexts.isEmpty {
+                    return nil
+                }
+
+                return selectedContexts.isEmpty ? String(localized: "Select all", bundle: .core)
+                                                : String(localized: "Deselect all", bundle: .core)
+            }
+            .assign(to: \.selectAllButtonTitle, on: self, ownership: .weak)
+            .store(in: &subscriptions)
+
+        interactor
+            .filterCountLimit
+            .map { limit -> String? in
+                switch limit {
+                case .base, .extended:
+                    return String(localized: "Select the calendars you want to see, up to \(limit.rawValue).", bundle: .core)
+                case .unlimited:
+                    return nil
+                }
+            }
+            .assign(to: \.filterLimitMessage, on: self, ownership: .weak)
+            .store(in: &subscriptions)
+
+        interactor
+            .filters
+            .sink { [weak self] filters in
+                guard let self else { return }
+
+                userFilter = filters.first { $0.context.contextType == .user }
+                courseFilters = filters
+                    .filter { $0.context.contextType == .course }
+                    .sorted()
+                groupFilters = filters
+                    .filter { $0.context.contextType == .group }
+                    .sorted()
+            }
+            .store(in: &subscriptions)
+    }
+}
+
+private extension SnackBarViewModel {
+
+    func showFilterLimitReachedMessage(limit: Int) -> Future<Void, Never> {
+        Future { [weak self] promise in
+            let message = String(localized: "You can only select up to \(limit) calendars.", bundle: .core)
+            self?.showSnack(message)
```

Comment:
It may be nice to not enqueue these snacks. I mean only show it if it's not already visible.
User may tap repeatedly on calendars to "try to make it work" (at least I did :) ) and getting multiple repeated messages after a some seconds of delay each time feels strange.
I think just checking `visibleSnack` equality might be enough.

Reply:
Good idea. I implemented this in the SnackBarViewModel.

Reply:
I also made the timer reset when a duplicate snack arrives.